### PR TITLE
save draft text

### PR DESCRIPTION
**Pull request by diablourbano**
This PR implements logic to save draft for chat texts under following circumstances:

- user wants to change from screen (similar to whatsapp)
- user needs/wants to change app
- user receives a call

The `draft` will be stored until the user clear it or send it.

### CODE REVIEW

File: src/actions/chatActions.js [hunk 1]
```diff
@@ -143,6 +149,49 @@ export const sendMessageByContactAction = (username: string, message: Object) =>
...
 };
 
+export const getChatDraftByContactAction = (contactId: string) => {
+  return async (dispatch: Function) => {
+    try {
+      const { drafts } = await storage.get('chat');
+      const [chatDraft, chatDrafts] = partition(drafts, { contactId });
+      const { draftText = '' } = chatDraft[0] || {};
+
+      if (!draftText || draftText === '') { return; }
```
Comment (jsidorenko) [inline src/actions/chatActions.js]:
```suggestion
      if (!draftText) return;
```

Comment (jsidorenko) [reply src/actions/chatActions.js]: `!draftText` covers empty strings as well

File: src/actions/chatActions.js [hunk 2]
```diff
@@ -143,6 +149,49 @@ export const sendMessageByContactAction = (username: string, message: Object) =>
   };
 };
 
+export const getChatDraftByContactAction = (contactId: string) => {
+  return async (dispatch: Function) => {
+    try {
+      const { drafts } = await storage.get('chat');
```
Comment (jsidorenko) [inline src/actions/chatActions.js]: I think if we put here `const { drafts = [] }` then we could avoid wrapping everything into try..catch.
What do you think?

File: src/actions/chatActions.js [hunk 3]
```diff
@@ -143,6 +149,49 @@ export const sendMessageByContactAction = (username: string, message: Object) =>
...
+  return async (dispatch: Function) => {
+    dispatch({
+      type: CLEAR_CHAT_DRAFT,
+    });
+  };
+};
+
+export const saveDraftAction = (contactId: string, draftText: string) => {
+  return async (dispatch: Function) => {
+    try {
```
Comment (jsidorenko) [inline src/actions/chatActions.js]: I'm wondering if Sentry would be able to catch the error when we wrapped everything into try..catch?

File: src/screens/Chat/Chat.js [hunk 4]
```diff
@@ -409,6 +467,8 @@ class ChatScreen extends React.Component<Props, State> {
             </View>}
           {!this.state.isFetching &&
             <GiftedChat
+              text={chatText}
+              onInputTextChanged={(text) => this.updateChatInput(text)}
```
Comment (jsidorenko) [inline src/screens/Chat/Chat.js]:
```suggestion
              onInputTextChanged={this.updateChatInput}
```

File: src/actions/chatActions.js (hunk 2 above)
Comment (diablourbano) [reply src/actions/chatActions.js]: The reason for the try/catch is mostly for the `storage.get('chat')` part, if for some reason Storage crashes.

Setting a default value should be there, so thanks for noticing it.

Comment (jsidorenko) [reply src/actions/chatActions.js]: If for some reason Storage crashes we should probably block the app and notify the user...
If you think `storage.get` could crash, then it's better to wrap the code of this `.get()` method inside the storage.js file as we're using `storage.get()` in a lot of places across the app

Comment (jsidorenko) [reply src/actions/chatActions.js]: Btw Sentry wrapps the whole codebase with the try..catch by default

Comment (diablourbano) [reply src/actions/chatActions.js]: ok 👍

**Pull request merged**